stl/inc/memory

```diff
@@ -3079,15 +3079,21 @@ inline void* align(size_t _Bound, size_t _Size, void*& _Ptr, size_t& _Space) noe
     return _Ptr;
 }
 
-#if _HAS_CXX20 && 0
+#if _HAS_CXX20
 template <size_t _Nx, class _Ty>
-_NODISCARD constexpr _Ty* assume_aligned(_Ty* _Ptr) noexcept /* strengthened */ {
+_NODISCARD constexpr _Ty* assume_aligned(_Ty* const _Ptr) noexcept /* strengthened */ {
+#ifndef __clang__ // TRANSITION, DevCom-1190972
+    if (_STD is_constant_evaluated()) {
+        return _Ptr;
+    }
+#endif // ^^^ workaround ^^^
+
     // this enforces the requirement that _Nx be a power of two
-    __builtin_assume_aligned(_Ptr, _Nx);
+    (void) __builtin_assume_aligned(_Ptr, _Nx);

```

Comment:
Shouldn't return resulting poitner cast back to `_Ty` ?

Reply:
That's a good question; what type does the builtin return? There was a sentence in the Standard about compilers maybe caring about whether the specific value returned by `assume_aligned` is used; is the builtin's behavior with respect to that documented by anyone?

Reply:
in gcc documentation only, I assume

Reply:
https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html

Reply:
cast from 'void *' is not allowed in a constant expression

or you mean:
 if constant expression return original pointer
if ordinal function call return static_cast<_Ty*>(__builtin_assume_aligned(_Ptr, _Nx))

Reply:
@StephanTLavavej >what type does the builtin return?

void*

https://imgur.com/a/qmrmIQI

Reply:
I'd want `static_cast<_Ty*>`.
Since it is not clear if the parameter of intrinsic is also assumed to be aligned, I'm for using the result

Reply:
MSVC does not care, but I'm uncertain about Clang. Clang documents that this intrinsic is implemented with GCC's semantics and the GCC manual is unclear.

TLDR we should probably use the return value of the intrinsic in non-constant-expressions. Alignment is meaningless in every `constexpr` evaluator I'm aware of so we may as well:
```c++
if (_STD is_constant_evaluated()) {
  return _Ptr;
} else {
  return static_cast<_Ty*>(__builtin_assume_aligned(_Ptr, N));
}
```
and ignore that MSVC and EDG reject calls to the intrinsic in constant expressions.

Reply:
What about https://developercommunity.visualstudio.com/content/problem/1190972/-builtin-assume-aligned-is-not-constexpr.html ?
Should I close it?

Reply:
Since the `static_cast` is necessary, and `constexpr`-incompatible, it seems reasonable to close the bug report - even if the compiler made the builtin constexpr, it wouldn't help anything.

Reply:
Done, I closed a bug in the https://developercommunity.visualstudio.com

And I filed a bug in gcc bugzilla. (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=97132)

https://gcc.godbolt.org/z/KK6W8G

They use static_cast without is_constant_evaluated.